### PR TITLE
SMTP Network Filter in contrib extensions

### PR DESCRIPTION
**Pull request by VishalDamgude**
Signed-off-by: VishalDamgude <vishal.damgude@freshworks.com>

Commit Message: Smtp Network Filter in contrib
Additional Description: 
This smtp network filter has been developed as per RFC 5321 (https://www.rfc-editor.org/rfc/rfc5321) and supports STARTTLS extension. It will parse SMTP commands and responses as per the SMTP protocol and provide protocol specific metrics.
The downstream starttls is supported by default. When smtp client sends STARTTLS, the filter will start secure transport on downstream connection. The upstream encryption is configurable. If enabled and if TLS handshake fails during upstream TLS negotiation, the filter responds to the client with error and closes the connection. If upstream TLS handshake succeeds, both the downstream and upstream connections are encrypted for secure smtp mail transfer.

Risk Level: Low
Testing: Unit Tests, Manual Testing
Docs Changes:
Release Notes:
Platform Specific Features: No
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/24219
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by VishalDamgude** [comment]
Hi @cpakulski @wbpcode ,
Created a new PR for SMTP Filter.Please check.

**Comment by wbpcode** [comment]
This is a super huge PR. Could you take a look when you have free time? Thanks. @cpakulski

**Comment by cpakulski** [comment]
@wpcode. Sure I can review it. The code seems to be based on my postgres filter, so constructs are quite familiar. Do you know how to enable/approve CI workflows?

**Comment by wbpcode** [comment]
Ci is approved now. The maintainers could approve it.

**Comment by phlax** [comment]
the approval is only required for the gh actions (mostly mobile tests) so it should not matter too much

seems like this PR is suffering some other azp issue, which i have seen from time to time, i will trigger manually

**Comment by phlax** [comment]
@VishalDamgude this is currently failing format checks https://dev.azure.com/cncf/envoy/_build/results?buildId=128650&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=33b57bf9-ef13-58c8-bc2b-f804aeb16d39

**Comment by VishalDamgude** [comment]
@phlax I have fixed the format error. Can you trigger the required checks again?

**Comment by phlax** [comment]
> Can you trigger the required checks again?

done. not sure why this is necessary - every so often azp decides it doesnt like a PR

**Comment by phlax** [comment]
@VishalDamgude copying out the CI advice

```
  Please fix your editor to ensure:

      - no trailing whitespace
      - no preceding mixed tabs/spaces
      - all files end with a newline
```

there are some other issues but best not ignore that advice

**Comment by cpakulski** [comment]
@plax Is it possible to give CI approval once, not each time a new code is pushed?

**Comment by VishalDamgude** [comment]
Hey @plax , we need CI approval again.

**Comment by phlax** [comment]
> Is it possible to give CI approval once, not each time a new code is pushed?

i dont think so - but until the PR is ready to land (unless it is testing mobile code) it probably doesnt matter much

**Comment by phlax** [comment]
apologies i see azp was not running again - ive kicked it - but its failing docs build

**Comment by phlax** [comment]
@VishalDamgude i think you need:

```diff
diff --git a/api/BUILD b/api/BUILD
index 5743322cc9..bdc45e3c32 100644
--- a/api/BUILD
+++ b/api/BUILD
@@ -91,6 +91,7 @@ proto_library(
         "//contrib/envoy/extensions/filters/network/sip_proxy/router/v3alpha:pkg",
         "//contrib/envoy/extensions/filters/network/sip_proxy/tra/v3alpha:pkg",
         "//contrib/envoy/extensions/filters/network/sip_proxy/v3alpha:pkg",
+        "//contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha:pkg",
         "//contrib/envoy/extensions/matching/input_matchers/hyperscan/v3alpha:pkg",
         "//contrib/envoy/extensions/network/connection_balance/dlb/v3alpha:pkg",
         "//contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha:pkg",
```

i tried to build the docs with that and it still failed as there are links to non-existent config docs pages

the docs can build by commenting those out:

```diff
diff --git a/api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto b/api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto
index 7cd5979f79..897045616d 100644
--- a/api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto
+++ b/api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto
@@ -13,8 +13,8 @@ option (udpa.annotations.file_status).work_in_progress = true;
 option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: SMTP Proxy]
-// SMTP Proxy :ref:`configuration overview
-// <config_network_filters_smtp_proxy>`.
+// [#comment: SMTP Proxy :ref:`configuration overview
+// <config_network_filters_smtp_proxy>`.]
 // [#extension: envoy.filters.network.smtp_proxy]
 
 message SmtpProxy {
@@ -28,8 +28,8 @@ message SmtpProxy {
     REQUIRE = 1;
   }
 
-  // The human readable prefix to use when emitting :ref:`statistics
-  // <config_network_filters_smtp_proxy_stats>`.
+  // [#comment: The human readable prefix to use when emitting :ref:`statistics
+  // <config_network_filters_smtp_proxy_stats>`].
   string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
 
   // Controls whether to establish upstream TLS connection to the server.

```

but i think we do need some docs before this lands

/docs

**Comment by VishalDamgude** [comment]
Hey @phlax , I've added docs. Can we trigger CI again?

**Comment by phlax** [comment]
im trying to understand why this PR needs to be kicked manually - can you merge main its pretty far behind current main

**Comment by phlax** [comment]
wierd, azp is still ignoring this PR

**Comment by VishalDamgude** [comment]
WIP.

File: contrib/smtp_proxy/filters/network/source/smtp_session.h
Comment (cpakulski) [inline contrib/smtp_proxy/filters/network/source/smtp_session.h:16]: Change enums to camel case. For enums it should start with capital letter and follow with small letters. Like:
ConnectionRequest
ConnectionSuccess
etc.

File: contrib/smtp_proxy/filters/network/source/smtp_session.h
Comment (cpakulski) [inline contrib/smtp_proxy/filters/network/source/smtp_session.h:30]: Change to _setTransactionState_. Function names should start with small letter.

**Review by cpakulski — CHANGES_REQUESTED** [review]
Please change coding style as indicated.

File: contrib/smtp_proxy/filters/network/source/smtp_session.h
Comment (VishalDamgude) [reply contrib/smtp_proxy/filters/network/source/smtp_session.h]: Done

File: contrib/smtp_proxy/filters/network/source/smtp_session.h
Comment (VishalDamgude) [reply contrib/smtp_proxy/filters/network/source/smtp_session.h]: Done

**Comment by VishalDamgude** [comment]
Can we start CI for this again @phlax ?

**Comment by VishalDamgude** [comment]
Hi @phlax,
I created a new PR https://github.com/envoyproxy/envoy/pull/26481 as suggested.
But CI is still not started. Can we check whats the issue?

**Comment by phlax** [comment]
>  But CI is still not started. Can we check whats the issue?

thanks - kicking the gh action CI on both to see if its related and triggers azp (again)

not sure on the problem - could be related to first-time contribution altho i see this from time-to-time on other PRs

i guess it may be something to do with the merge/fork algo that is confused with this set of changes specifically (where forked from originally perhaps)

**Comment by lizan** [comment]
Closing in favor of https://github.com/envoyproxy/envoy/pull/26481

**Pull request closed without merging**
